File: src/main/java/WebviewCallbackWrapper.java
import co.casterlabs.rakurai.json.Rson;
import co.casterlabs.rakurai.json.element.JsonArray;
import co.casterlabs.rakurai.json.element.JsonElement;
import dev.webview.ConsumingProducer;

public final class WebviewCallbackWrapper {
    private static final Rson r = new Rson.Builder().build();

    public static ConsumingProducer<JsonArray, JsonElement> wrap(Handler handler) {
        return jsonArray -> {
            try {
                String jsonArrayAsString = String.valueOf(jsonArray);
                Object result = handler.callback.apply(jsonArrayAsString);
                return r.toJson(result);
            } catch (Exception e) {
                System.out.println("Failed to execute " + handler.name + ". Returned `null` instead.");
                System.out.println("Error: " + e.getMessage());
                return null;
            }
        };
    }
}

File: src/main/java/WebviewCallbackWrapper.kt
import co.casterlabs.rakurai.json.Rson
import co.casterlabs.rakurai.json.element.JsonArray
import co.casterlabs.rakurai.json.element.JsonElement
import dev.webview.ConsumingProducer
import org.slf4j.LoggerFactory

/**
 * Represents a handler that is called from the frontend.
 */
data class Handler(var name: String, var callback: (String) -> Any?)

/**
 * The current implementation of the `webview` package requires that methods passed to
 * `webview.bind()` receive a JsonArray and return JsonElement of the casterlabs package.
 * So this is a wrapper to make things more intuitive.
 */
object WebviewCallbackWrapper {
    private val r: Rson = Rson.Builder().build()
    private val log = LoggerFactory.getLogger(this::class.java)

    /**
     *
     * @param handler a method that receives a json string of the arguments from the frontend,
     * and returns some java object in return.
     * @return a valid method for `webview.bind()` that essentially does what `handler()` does.
     */
    fun wrap(handler: Handler) =
        ConsumingProducer { jsonArray: JsonArray? ->
            try {
                r.toJson<Any>(handler.callback(jsonArray.toString()))
            } catch (e: Exception) {
                log.error("Failed to execute ${handler.name}. Returned `null` instead.", e)
                null
            }
        }
}